### DISCUSSION TITLE
Prescaled TCAn clock doesn't reset when Event CNTB reset is used

### DISCUSSION THREAD

**Discussion by SimonMerrett** [SpenceKonde/megaTinyCore#750] [Q&A]
Not sure if this is Q or A but I thought I'd post an observation here, in case it is helpful or someone spots how I'm using the TCA peripheral "wrong".

So like much of my recent activity on ATtiny3227, I'm attempting to make an arbitratry clock divider. That is, it could take a 10 MHz input and divide it by 10 million to give 1 pulse per second, or it could divide by 9,999,999 to give slightly more than 1 pulse per second etc.

I had been initially trying to use TCB1, fed with prescaled RTC (and everything running from 10 MHz extclk) to give me the top 16 MSBs of a 24 bit counter and then running TCB0 as a single shot to provide the remaining LSB count. There were two problems with this: my event pulse at completion of TCB0 was too short as a useful 1PPS signal and instead of restarting at the end of TCB0, TCB1 would be merrily beginning again/ resetting and counting in sync with prescaled RTC as soon as TCB0 had begun its count, rather than waiting until the end. 

To address this, I removed prescaled RTC and TCB1 and replaced them with prescaled TCA0 to provide the 16 MSBs and kept TCB0 single shot for the 8 LSBs, triggering on TCA0 PER event. Then I used TCA0 CNTB to reset TCA0 from the TCB0 CCMP event. This left TCB1 for my longer 1 pulse per second generation. 

Initially this showed promise. To check everything was working as expected, I reduced the TCA0 prescaler, TCA0 PER and TCB0 CCMP so that multiple cycles would fit on the oscilloscope screen with external clock pulses visible for counting. From my initial testing, I am concluding that TCA0 CNTB reset (on rising edge of event) results in a CNT reset but the prescaled clock is unaffected, meaning that for e.g. a prescaler of 256 the event could arrive/reset TCA0 when at 64/256 clock cycles and the first edge/count could be 90 degrees out of phase with your CNTB event edge.

The datasheet diagram suggests this is expected, although it is disappointing because it reduces the utility of the prescaled TCA0 for increasing the 16 bit timer's range to 24 and 32 bit utility in conjunction with other timers.
![image](https://user-images.githubusercontent.com/26767525/181238492-1f915e5a-2e78-46b7-8e26-2bcb01d4a41f.png)

If anyone is curious in the near term, I can set up some screen shots and post code but I thought just a description would be helpful to flag this observed behaviour. Looks like I need an DB series to give me enough timers to generate my initial prescaler clock (that can be reset when my single shot LSB timer fires, my MSB clock to be driven from the prescaler clock and another one to generate the 1 pulse per second in a responsive way that lasts long enough. I may try playing around some more with ATtiny3227 but I don't think I can see a way to generate my longer pulse, even if I do get it starting at the right time.

**Comment by SpenceKonde** [comment]
Hmm, and syncupd on TCB does it in the other direction it sounds like? 

Doesn't the TCA prescaler reset when the peripheral is disabled + enabled (hereafter, "bounced"), so you could bounce the TCA in in an interrupt while TCB was outputting the LSBs? 
(the interrupt would begin executing 6 clocks after the thing that triggered it; assuming the prescaler value for TCA is constant, the compiler generated ISR for TCA0.SINGLE.CTRLA = 0; TCA0.SINGLE.CTRLA=N;) would be:
```
push r0 //1
push r1 //1
in r0, SREG //1
push r0 // 1
eor r1, r1 //1  - 5 so far
push SomeUpperRegister //2
sts TCA.SINGLE.CTRLA, r1 //2
ldi SomeUpperRegister N //1 - another 5
sts TCA.SINGLE.CTRLA, SomeUpperRegister //2
pop SomeUpperRegister //2
pop r0 //2
out SREG, r0 //1
pop r1 //2
pop r0 // 2
reti //4 - annother 15
```
for total interrupt time of 6 + 25 = 31 clocks. with timer restarted on clock 12 of isr 18 clocks after the thing that triggered it, and 13 clocks before normal code execution resumes. 

Hand written NAKED_ISR containing inline assembly could be 
```
// Only need one register, but r0 and r1 are useless because we need to ldi into it so don't fuck with those. 
// and we are not performing any mathematical or logical operations so we don't need to bother saving SREG
// since none of these instructions can change it. 
push r16; //1
ldi r16, 0 //1
sts TCA.SINGLE.CTRLA, r16 //1
ldi r16, N //1
sts TCA.SINGLE.CTRLA, r16 //2 
pop r16; //2
reti; // 4
```
6 + 12 = 18 clocks total  with clock restarted on clock 6 of isr 12 clocks after isr triggered and 6 clocks before normal code execution resumes 



What exactly are the requirements here for maximum period and accuracy here? I'm kind of foggy on what you're trying to do, but I'll bet you could make something happen with the CCLs and TCBs, even if you can't bounce the TCA to reset it's prescaler (using TCA for the 1 PPS and the two TCBs to do this pulse generation). Though resetting a CCL prescaler still takes bouncing the CCL peripheral.

**Reply by SimonMerrett** [reply]
This afternoon think I've managed to crack the first part - 1 PPS from 10 MHz extclk with a good response time for the rising pulse edge and  helpful duty on the pulse (e.g. more useful than 1 clock cycle that you get on the event pulses). Will post it here when I've tidied it up, and would appreciate a critique. Indeed, I did use the CCL, as well as all three TCs.

**Reply by SpenceKonde** [reply]
Prescaling with a CCD, in case it's relevant: 

First column is division factor, LUTS and Events is the number of LUT and event channels it takes. Only the least resource intensive way to get a given divider is shown. 

| Total | p0 | 0 | 1 | oppr 0    | oppr 1    | p2+3 | p2 | 2 | 3 | oppr 2    | oppr 3    | p3 | 4 | 5 | op4 | LUTS | Events |
|-------|----|---|---|-----------|-----------|------|----|---|---|-----------|-----------|----|---|---|-----|------|--------|
|     4 |  4 | 4 | - |         - |         - |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|     8 |  8 | 4 | - |         - |         - |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|    12 | 12 | 4 | 8 |       ADD |         - |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|    16 | 16 | 8 | 8 |       ADD |         - |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|    20 | 16 | 4 | 4 |      ADD2 |       ADD |    - |  4 | 4 | - |       ADD |         - |  - | - | - |   - |    3 |      1 |
|    24 | 16 | 4 | 4 |      ADD2 |       ADD |    - |  8 | 8 | - |       ADD |         - |  - | - | - |   - |    3 |      1 |
|    28 |  8 | 4 | 4 |       ADD |       ADD |   20 |  8 | 4 | 4 |       ADD |       ADD | 12 | 4 | 8 | ADD |    6 |      0 |
|    32 | 32 | 4 | 8 |       MUL |           |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|    36 |  8 | 4 | 4 |       ADD |       ADD |   28 | 12 | 4 | 8 |       ADD |       ADD | 16 | 8 | 8 | ADD |    6 |      0 |
|    40 |  8 | 4 | 4 |       ADD |       ADD |   32 | 16 | 8 | 8 |       ADD |       ADD | 16 | 8 | 8 | ADD |    6 |      0 |
|    44 | 12 | 4 | 8 |       ADD |  ADD2MUL3 |    - |  - | 4 | 8 |       ADD |         - |  - | - | - |   - |    4 |      2 |
|    48 | 16 | 4 | 4 |      ADD2 |       ADD |    - | 32 | 4 | - |       ADD |  ADD4MUL5 |  - | 4 | 4 | ADD |    5 |      1 |
|    52 | 32 | 4 | 8 |       MUL |       ADD |    - | 16 | 4 | 4 |       MUL |      ADD4 |  - | 4 | - |   - |    5 |      1 |
|    64 | 64 | 8 | 8 |       MUL |           |    - |  - | - | - |         - |         - |  - | - | - |   - |    2 |      0 |
|    76 | 12 | 4 | 8 |       ADD |       ADD |   64 |  8 | 4 | 4 |       ADD |       MUL |  8 | 4 | 4 | ADD |    6 |      0 |
|    80 | 16 | 8 | 8 |       ADD |  ADD2MUL3 |    - |  - | 8 | 8 |       ADD |         - |  - | - | - |   - |    4 |      2 |
|    84 | 64 | 8 | 8 |       MUL |       ADD |    - | 16 | 8 | 8 |       ADD |      ADD4 |  - | 4 | - |   - |    5 |      1 |
|    88 |  8 | 4 | 4 |       ADD |  ADD2MUL3 |    - |  - | 4 | 4 |       ADD |  ADD4MUL5 |  - | 4 | 4 | ADD |    6 |      1 |
|    96 |  8 | 4 | 4 |       ADD |       ADD |    - | 12 | 4 | 8 |       ADD |      MUL4 |  - | 4 | - |   - |    5 |      1 |
|   100 | 64 | 8 | 8 |       MUL |       ADD |    - | 32 | 4 | 8 |       MUL |      ADD4 |  - | 4 | - |   - |    5 |      1 |
|   108 | 12 | 4 | 8 |       ADD |  ADD2MUL3 |    - |  - | 4 | 8 |       ADD |  ADD4MUL5 |  - | 4 | 4 | ADD |    6 |      1 |
|   128 | 16 | 4 | 4 |      MUL2 |       MUL |    - |  8 | 8 | 1 |       MUL |         - |  - | - | - |   - |    3 |      1 |
|   140 | 12 | 4 | 8 |       ADD |       ADD |  128 | 16 | 8 | 8 |       ADD |       MUL |  8 | 4 | 4 | ADD |    6 |      0 |
|   144 | 12 | 4 | 8 |       ADD |       MUL |    - | 12 | 4 | 8 |       ADD |         - |  - | - | - |   - |    4 |      1 |
|   152 |  8 | 4 | 4 |       ADD |  ADD2MUL3 |    - |  - | 4 | 4 |       ADD |  ADD4MUL5 |  - | 4 | 8 | ADD |    6 |      1 |
|   160 |  8 | 4 | 4 |       ADD |       MUL |    - | 12 | 4 | 8 |       ADD |      ADD4 |  - | 8 | - |   - |    5 |      1 |
|   172 | 12 | 4 | 8 |       ADD |  ADD2MUL3 |    - |  - | 4 | 8 |       ADD |  ADD4MUL5 |  - | 4 | 8 | ADD |    6 |      1 |
|   192 | 16 | 8 | 8 |       ADD |       MUL |    - | 12 | 4 | 8 |       ADD |         - |  - | - | - |   - |    4 |      1 |
|   204 | 12 | 4 | 8 |       ADD |       ADD |  192 | 16 | 8 | 8 |       ADD |       MUL | 12 | 4 | 8 | ADD |    6 |      0 |
|   208 | 16 | 8 | 8 |       ADD |  ADD2MUL3 |    - |  - | 8 | 8 |       ADD |  ADD4MUL5 |  - | 4 | 4 | ADD |    6 |      1 |
|   224 | 32 | 4 | 8 |       MUL |       ADD |  192 | 16 | 8 | 8 |       ADD |       MUL | 12 | 4 | 8 | ADD |    6 |      0 |
|   256 | 16 | 4 | 4 |       MUL |       MUL |    - | 16 | 4 | 4 |       MUL |         - |  - | - | - |   - |    4 |      1 |
|   268 | 12 | 4 | 8 |       ADD |       ADD |  256 | 16 | 8 | 8 |       ADD |       MUL | 16 | 8 | 8 | ADD |    6 |      0 |
|   288 | 12 | 4 | 8 |       ADD |       MUL |    - | 16 | 8 | 8 |       ADD |      ADD4 |  - | 8 | - |   - |    5 |      1 |
|   300 | 12 | 4 | 8 |       ADD |  ADD2MUL3 |    - |  - | 4 | 8 |       ADD |  ADD4MUL5 |  - | 8 | 8 | ADD |    6 |      1 |
|   336 | 16 | 8 | 8 |       ADD |  ADD2MUL3 |    - |  - | 8 | 8 |       ADD |  ADD4MUL5 |  - | 4 | 8 | ADD |    6 |      1 |
|   384 | 16 | 8 | 8 |       ADD |       MUL |    - | 16 | 4 | 4 |       MUL |      ADD4 |  - | 8 | - |   - |    5 |      1 |
|   512 | 64 | 8 | 8 |       MUL |       ADD |    - | 16 | 8 | 8 |       ADD |      MUL4 |  - | 4 | - |   - |    5 |      1 |
|   576 | 12 | 4 | 8 |       ADD |       MUL |   48 | 32 | 4 | 8 |       MUL |       ADD | 16 | 4 | 4 | MUL |    6 |      0 |
|   592 | 16 | 8 | 8 |       ADD |  ADD2MUL3 |    - |  - | 8 | 8 |       ADD |  ADD4MUL5 |  - | 8 | 8 | ADD |    6 |      1 |
|   640 | 16 | 4 | 4 |       MUL |       MUL |    - | 32 | 4 | 8 |       MUL |      ADD4 |  - | 8 | - |   - |    5 |      1 |
|   768 | 12 | 4 | 8 |       ADD |       MUL |   64 | 32 | 4 | 8 |       MUL |       ADD | 32 | 4 | 8 | MUL |    6 |      0 |
|  1024 | 32 | 4 | 8 |       MUL |       MUL |    - | 32 | 4 | 8 |       MUL |         - |  - | - | - |   - |    4 |      1 |
|  1088 | 16 | 4 | 4 |      MUL2 |       MUL |    - | 68 | 8 | 1 |       MUL |  ADD4MUL5 |  - | 8 | 4 | MUL |    5 |      1 |
|  1152 | 12 | 4 | 8 |       ADD |       MUL |    - | 16 | 8 | 8 |       ADD |      MUL4 |  - | 8 | - |   - |    5 |      1 |
|  1280 | 16 | 4 | 4 |       MUL |       MUL |   80 | 16 | 4 | 4 |       MUL |       ADD | 64 | 8 | 8 | MUL |    6 |      0 |
|  1536 | 16 | 8 | 8 |       ADD |       MUL |    - | 16 | 4 | 4 |       MUL |      MUL4 |  - | 8 | - |   - |    5 |      1 |
|  2048 | 16 | 4 | 4 |       MUL |       MUL |    - | 32 | 4 | 8 |       MUL |      MUL4 |  - | 8 | - |   - |    5 |      1 |
|  2304 | 32 | 4 | 8 |       MUL |       MUL |    - | 64 | 8 | 8 |       MUL |      ADD4 |  - | 8 | - |   - |    5 |      1 |
|  4096 | 64 | 8 | 8 |       MUL |       MUL |    - | 64 | 8 | 8 |       MUL |         - |  - | - | - |   - |    4 |      1 |
|  4608 | 64 | 8 | 8 |       MUL |       MUL |    - | 64 | 8 | 8 |       MUL |      ADD4 |  - | 8 | 4 | ADD |    5 |      1 |
|  8192 | 32 | 4 | 8 |       MUL |       MUL |    - | 64 | 8 | 8 |       MUL |      MUL4 |  - | 8 | - |   - |    5 |      1 |
| 16384 | 16 | 8 | 8 |       ADD |       MUL | 1024 | 32 | 4 | 8 |       MUL |       MUL | 32 | 4 | 8 | MUL |    6 |      0 |
| 24576 | 12 | 4 | 8 |       ADD |       MUL | 2048 | 32 | 4 | 8 |       MUL |       MUL | 64 | 8 | 8 | MUL |    6 |      0 |
| 32768 | 64 | 8 | 8 |       MUL |       MUL |    - | 64 | 8 | 8 |       MUL |      MUL4 |  - | 8 | 4 | ADD |    5 |      1 |
| 65536 | 16 | 8 | 8 |       ADD |       MUL | 4096 | 64 | 8 | 8 |       MUL |       MUL | 64 | 8 | 8 | MUL |    6 |      0 |
|131072 | 64 | 8 | 8 |       MUL |       MUL | 2048 | 32 | 4 | 8 |       MUL |       MUL | 64 | 8 | 8 | MUL |    6 |      0 |
|262144 | 64 | 8 | 8 |       MUL |       MUL | 4096 | 64 | 8 | 8 |       MUL |       MUL | 64 | 8 | 8 | MUL |    6 |      0 |

**Reply by SimonMerrett** [reply]
This is brilliant. I'm afraid I needed larger divisors but this is the likely to be helpful for other uses. I posted my working code in a different discussion as it circumvents the behaviour I reported above.

**Comment by SpenceKonde** [comment]
Check the errata. I think there is a bug there

**Reply by SimonMerrett** [reply]
I couldn't see anything in https://ww1.microchip.com/downloads/aemDocuments/documents/MCU08/ProductDocuments/Errata/ATtiny3224-3226-3227-SilConErrataClarif-DS40002342A.pdf
